File: packages/shared/src/form/callingCodes.ts
/**
 * A small, generic ISO-3166 + calling-code table bundled with the builder.
 *
 * This is the *default* countryCode/callingCode dropdown source, used whenever the
 * builder user hasn't selected a Samsung subsidiary code in the Configure step. The
 * reference implementation's Samsung-org-specific `country_subsidiary`/`subsidiary_detail`
 * tables (`Final_forms_format/SGE-EN_F2H26.js`) are transcribed separately in
 * `subsidiaryData.ts` and take over the dropdown contents once a subsidiary is selected
 * — see `codegen/js/buildDataJs.ts`.
 *
 * `mobileDigits` also replaces the reference's hardcoded `mobileNumberByCountry`
 * validator, which special-cased UAE=9 digits vs. everything else=8 digits — that rule
 * is made data-driven per country here instead, and is reused as a lookup (with the
 * reference's UAE=9/else=8 rule as fallback) for subsidiary-driven entries too.
 */
export interface CallingCodeEntry {
  countryCode: string; // ISO-3166 alpha-2
  callingCode: string; // without leading "+"
  countryName: string;
  mobileDigits: number; // expected local mobile-number digit count (excluding calling code)
}

export const CALLING_CODES: readonly CallingCodeEntry[] = [
  { countryCode: "AE", callingCode: "971", countryName: "United Arab Emirates", mobileDigits: 9 },
  { countryCode: "SA", callingCode: "966", countryName: "Saudi Arabia", mobileDigits: 9 },
  { countryCode: "EG", callingCode: "20", countryName: "Egypt", mobileDigits: 10 },
  { countryCode: "IL", callingCode: "972", countryName: "Israel", mobileDigits: 9 },
  { countryCode: "PS", callingCode: "970", countryName: "Palestine", mobileDigits: 9 },
  { countryCode: "TR", callingCode: "90", countryName: "Turkey", mobileDigits: 10 },
  { countryCode: "JO", callingCode: "962", countryName: "Jordan", mobileDigits: 9 },
  { countryCode: "LB", callingCode: "961", countryName: "Lebanon", mobileDigits: 8 },
  { countryCode: "IQ", callingCode: "964", countryName: "Iraq", mobileDigits: 10 },
  { countryCode: "KW", callingCode: "965", countryName: "Kuwait", mobileDigits: 8 },
  { countryCode: "QA", callingCode: "974", countryName: "Qatar", mobileDigits: 8 },
  { countryCode: "BH", callingCode: "973", countryName: "Bahrain", mobileDigits: 8 },
  { countryCode: "OM", callingCode: "968", countryName: "Oman", mobileDigits: 8 },
  { countryCode: "GB", callingCode: "44", countryName: "United Kingdom", mobileDigits: 10 },
  { countryCode: "US", callingCode: "1", countryName: "United States", mobileDigits: 10 },
  { countryCode: "FR", callingCode: "33", countryName: "France", mobileDigits: 9 },
  { countryCode: "DE", callingCode: "49", countryName: "Germany", mobileDigits: 10 },
  { countryCode: "IN", callingCode: "91", countryName: "India", mobileDigits: 10 },
  { countryCode: "PK", callingCode: "92", countryName: "Pakistan", mobileDigits: 10 },
];

export function findCallingCodeEntry(countryCode: string): CallingCodeEntry | undefined {
  return CALLING_CODES.find((c) => c.countryCode === countryCode.toUpperCase());
}

File: backend-py/app/form_pipeline/form/calling_codes.py
"""Port of packages/shared/src/form/callingCodes.ts.

A small, generic ISO-3166 + calling-code table bundled with the builder. This
is the *default* countryCode/callingCode dropdown source, used whenever no
Samsung subsidiary code has been selected — see subsidiary_data.py for the
org-specific tables that take over once one is.
"""

from __future__ import annotations

from typing import NamedTuple, Optional


class CallingCodeEntry(NamedTuple):
    countryCode: str  # ISO-3166 alpha-2
    callingCode: str  # without leading "+"
    countryName: str
    mobileDigits: int  # expected local mobile-number digit count (excluding calling code)


CALLING_CODES: tuple[CallingCodeEntry, ...] = (
    CallingCodeEntry("AE", "971", "United Arab Emirates", 9),
    CallingCodeEntry("SA", "966", "Saudi Arabia", 9),
    CallingCodeEntry("EG", "20", "Egypt", 10),
    CallingCodeEntry("IL", "972", "Israel", 9),
    CallingCodeEntry("PS", "970", "Palestine", 9),
    CallingCodeEntry("TR", "90", "Turkey", 10),
    CallingCodeEntry("JO", "962", "Jordan", 9),
    CallingCodeEntry("LB", "961", "Lebanon", 8),
    CallingCodeEntry("IQ", "964", "Iraq", 10),
    CallingCodeEntry("KW", "965", "Kuwait", 8),
    CallingCodeEntry("QA", "974", "Qatar", 8),
    CallingCodeEntry("BH", "973", "Bahrain", 8),
    CallingCodeEntry("OM", "968", "Oman", 8),
    CallingCodeEntry("GB", "44", "United Kingdom", 10),
    CallingCodeEntry("US", "1", "United States", 10),
    CallingCodeEntry("FR", "33", "France", 9),
    CallingCodeEntry("DE", "49", "Germany", 10),
    CallingCodeEntry("IN", "91", "India", 10),
    CallingCodeEntry("PK", "92", "Pakistan", 10),
    CallingCodeEntry("IR", "98", "Iran", 10),
)


def find_calling_code_entry(country_code: str) -> Optional[CallingCodeEntry]:
    upper = country_code.upper()
    for entry in CALLING_CODES:
        if entry.countryCode == upper:
            return entry
    return None
